extension/apikeyauthextension/authenticator_test.go

```diff
@@ -477,6 +477,179 @@ func TestAuthenticator_AuthorizationHeader(t *testing.T) {
+				Application: "my_app",
+				Resources:   []string{"my-resource"}, // no format, just raw value
+				Privileges:  []string{"write"},
+			}},
+		}, body)
+		assert.NoError(t, json.NewEncoder(w).Encode(successfulResponse))
+	})
+
+	config := createDefaultConfig().(*Config)
+	config.ApplicationPrivileges = []ApplicationPrivilegesConfig{{
+		Application: "my_app",
+		Privileges:  []string{"write"},
+		DynamicResources: []DynamicResource{{
+			Metadata: "X-Resource-Name",
+			// Format empty, should default to "%s"
+		}},
+	}}
+	config.Cache.KeyMetadata = []string{"X-Resource-Name"}
+	authenticator := newTestAuthenticator(t, srv, config)
+
+	withMetadata := client.NewContext(context.Background(), client.Info{
+		Metadata: client.NewMetadata(map[string][]string{
+			"X-Resource-Name": {"my-resource"},
+		}),
+	})
+	ctx, err := authenticator.Authenticate(withMetadata, map[string][]string{
+		"Authorization": {"ApiKey " + base64.StdEncoding.EncodeToString([]byte("id:secret"))},
+	})
+	assert.NoError(t, err)
+	clientInfo := client.FromContext(ctx)
+	assert.Equal(t, user, clientInfo.Auth.GetAttribute("username"))
+}
+
+func TestAuthenticator_DynamicResourcesMissingMetadata(t *testing.T) {
+	config := createDefaultConfig().(*Config)
+	config.ApplicationPrivileges = []ApplicationPrivilegesConfig{{
+		Application: "my_app",
+		Privileges:  []string{"write"},
+		DynamicResources: []DynamicResource{{
+			Metadata: "X-Resource-Name",
+			Format:   "resource:%s",
+		}},
+	}}
+	config.Cache.KeyMetadata = []string{"X-Resource-Name"}
+	authenticator := newTestAuthenticator(t, newMockElasticsearch(t, newCannedHasPrivilegesHandler(successfulResponse)), config)
+
+	// Missing metadata should return an error, since it is required
+	// to be in the cache key, and that is computed before formatting
+	// the dynamic resource for checking application privileges.
+	_, err := authenticator.Authenticate(context.Background(), map[string][]string{
+		"Authorization": {"ApiKey " + base64.StdEncoding.EncodeToString([]byte("id:secret"))},
+	})
+	require.Error(t, err)
+	assert.EqualError(t, err, `error computing cache key: missing client metadata "X-Resource-Name"`)
+}
+
+func TestAuthenticator_DynamicResourcesMultipleMetadataValues(t *testing.T) {
+	config := createDefaultConfig().(*Config)
+	config.ApplicationPrivileges = []ApplicationPrivilegesConfig{{
+		Application: "my_app",
+		Privileges:  []string{"write"},
+		DynamicResources: []DynamicResource{{
+			Metadata: "X-Resource-Name",
+			Format:   "resource:%s",
+		}},
+	}}
+	config.Cache.KeyMetadata = []string{"X-Resource-Name"}
+	authenticator := newTestAuthenticator(t, newMockElasticsearch(t, newCannedHasPrivilegesHandler(successfulResponse)), config)
+
+	// Multiple metadata values should return an error
+	withMultipleValues := client.NewContext(context.Background(), client.Info{
+		Metadata: client.NewMetadata(map[string][]string{
+			"X-Resource-Name": {"resource1", "resource2"},
+		}),
+	})
```

Comment:
I need help to understand this part. Why is this an issue?

Reply:
Are you asking why we expect an error when there are multiple values for a header? Because otherwise how would we format the resource?

Reply:
Yes. From my understanding, we can have multiple resources, right? For example, in our elastic config in TF [we use a wildcard](https://github.com/elastic/obs-ds-managed-services-testing/blob/931d0ada48fa35f28f4524c7c655c99c98b4946d/terraform/elastic/serverless-project/main.tf#L28). So why would it be an issue if we define multiple here?

> Because otherwise how would we format the resource?

I would expect the formatter applies to all elements of the list